src/corecel/sys/KernelDiagnostics.hh

```diff
@@ -156,8 +166,8 @@ inline auto KernelDiagnostics::insert(F*           func,
     static_assert(std::is_function<F>::value,
                   "KernelDiagnostics must be called with a function object, "
                   "not a function pointer or anything else.");
-    auto iter_inserted = keys_.insert(
-        {reinterpret_cast<std::uintptr_t>(func), key_type{this->size()}});
+    auto iter_inserted
+        = this->threadsafe_insert_new(reinterpret_cast<std::uintptr_t>(func));
```

Comment:
Since this is only called once per kernel instantiation (assuming the kernels use the `CELER_LAUNCH_KERNEL` macro), I don't think we have to worry too much about the performance of the mutex inside this call.

Reply:
Sorry, I meant instantiation of `KernelParamsCalculator`: so once per program per kernel. Not once per kernel call.